File: go/encapsulate_conditionals.go
// Encapsulate Conditionals
// Extract complex conditionals into functions that convey the intent of the condition.
// Create names that reveal the intent of the conditional.

package main

import "fmt"

type User struct {
	Age       int
	Children  bool
	IsPremium bool
	Coupon    bool
}

func NewUser() *User {
	return &User{
		Age:       25,
		Children:  false,
		IsPremium: false,
		Coupon:    false,
	}
}

func (u *User) HasChildren() bool {
	return u.Children
}

func (u *User) PremiumMember() bool {
	return u.IsPremium
}

func (u *User) HasCoupon() bool {
	return u.Coupon
}

// Before
func checkAvailabilityBefore(user *User) {
	if user.Age > 18 && !user.HasChildren() && (user.PremiumMember() || user.HasCoupon()) {
		fmt.Println("User is available for the offer.")
	} else {
		fmt.Println("User is not available for the offer.")
	}
}

// After
func checkAvailability(user *User) {
	if isAvailable(user) {
		fmt.Println("User is available for the offer.")
	} else {
		fmt.Println("User is not available for the offer.")
	}
}

func isAvailable(user *User) bool {
	if user.Age <= 18 {
		return false
	}
	if user.HasChildren() {
		return false
	}
	if user.PremiumMember() {
		return true
	}
	if user.HasCoupon() {
		return true
	}
	return false
}

func main() {
	user := NewUser()
	
	checkAvailabilityBefore(user)
	fmt.Println()
	checkAvailability(user)
}

File: kotlin/encapsulate_conditionals.kt
// Encapsulate Conditionals
// Extract complex conditionals into functions that convey the intent of the condition.
// Create names that reveal the intent of the conditional.

class User {
    val age = 25
    private val children = false
    private val isPremium = false
    private val coupon = false

    fun hasChildren(): Boolean = children
    fun premiumMember(): Boolean = isPremium
    fun hasCoupon(): Boolean = coupon
}

// Before
fun checkAvailabilityBefore(user: User) {
    if (user.age > 18 && !user.hasChildren() && (user.premiumMember() || user.hasCoupon())) {
        println("User is available for the offer.")
    } else {
        println("User is not available for the offer.")
    }
}

// After
fun checkAvailability(user: User) {
    if (isAvailable(user)) {
        println("User is available for the offer.")
    } else {
        println("User is not available for the offer.")
    }
}

private fun isAvailable(user: User): Boolean {
    if (user.age <= 18) return false
    if (user.hasChildren()) return false
    if (user.premiumMember()) return true
    if (user.hasCoupon()) return true
    return false
}

fun main() {
    val user = User()
    
    checkAvailabilityBefore(user)
    println()
    checkAvailability(user)
}
